ENH,API: Make the errstate/extobj a contextvar

seberg opened the pull request:
~**This builds on gh-23922, so a draft.  Also needs some release notes, but it is ready for review (the diff is just too large without gh-23992)~**

This does a few changes:
* Clean up a bit (a few commits to simplify)
* Change the core to be a capsule which holds the bufsize.  Then moves parsing fully to C (the split doesn't make life easier anyway).  In short, it completely revamps most of the errstate handling internally.
* More cleanup, hide away all those Python and C constants nobody should see.

In the end, things are now well encapsulated into `extobj.c` rather than sprinkled around.  It 

The choice of a capsule is a bit :shrug: a custom object may actually be almost as lean and neat because it removes some of that custom cleanup.

End-user effects:
* `np.errstate` is now actually thread-safe *and* additionally context safe.
* `np.errstate` cannot be entered twice (but that doesn't make sense).  I made its attribute private also (but you shouldn't use them).
  * I do allow entering multiple times (if you exited first), because the random code somewhere did it.
* Things are faster.
* `np.errstate` is *much* faster (I think about 2.5-3 times), not sure if the `__slots__` matter, but why not.

----

There is still a bunch of follow-ups that should happen eventually.  The way the ufuncs use this API isn't adapted yet (and passes unused stuff maybe even, but maybe will add that part here).  Mainly, we probably should get the `extobj` exactly once.


Closes gh-9444
Closes gh-19006

seberg committed:
MAINT: Refactor away helper and remove errstate default optimization

The optimization doesn't make sense and is incorrect to begin with.
It just stands in the way of moving to a contextvar.
(Maybe there is a way to optimize the contextvar access, but it has
to look differently and should be done after we got it.)

seberg committed:
MAINT: Untangle error callable `pyfunc` and ufunc name

Mainly, pass the ufunc name everywhere explicitly, rather than
building an unnecessary tuple.

seberg committed:
MAINT: Remove unused `first` handling in ufunc exception raising

seberg committed:
ENH: Working restructure of the errstate/extobj into a contextvar

seberg committed:
MAINT: Unceremonously remove errobj/extobj related constants

Move them to live exclusively in `extobj.c` source file where they
are actually used.

seberg committed:
MAINT: Silence compiler warning about uninitialized value

(I don't think its correct, but OK)

seberg commented:
PyPy doesn't support `PyContextVar_Resets()`?  @mattip I guess calling back into python is probably be the best choice (with an `#ifdef` guard)?  Or I can expose the contextvar to Python, but I somewhat like it hidden away.
(emscript test is already fixed/skipped on other PR.)

mattip commented:
Yay, less code and faster too! Does the speedup show up in the benchmarks?

You are using a capsule because it is easier than a typeobject? I guess that is fine, and it is an implementation detail we can change later.

> PyPy doesn't support `PyContextVar_Reset()`?

No, as of now it does not. Basically PyPy implements all the contextvar C-API by calling back into the interpreter's `_contextvars` module, which in PyPy is implemented in pure python following the code in PEP 567. In fact, on PyPy it would be preferable to do as much of this in pure python as possible. Is there a reason you decided to implement this in C rather than in python?

seberg commented:
> Is there a reason you decided to implement this in C rather than in python?

I thought I would hide things, and changing the contextvar in C seemed fair (so that `_seterrobj` does both.  It is arguably simpler (and even a bit faster).

I just did a change to expose it and at least do the `contextvar.reset(token)` in python.  I need to create the capsule in C.  Would it be better for PyPy if `_seterrobj()` just returned the capsule and we add it to the contextvar in Python?  (i.e. switches between C/Python one less time?)

I think that would be fine also.

On the capsule:  yes, its just slightly easier to implement, although means two objects are created (capsule struct + capsule).  This is one of those things where cython would be nice (also since we could move `np.errstate()` to cython maybe.  But I think its a better followup.

seberg committed:
MAINT: Expose `_extobj_contextvar` to python and reset from there

This fixes PyPy, which doesn't implement the full C-side API.
It is also actually faster probably...

mattip commented:
> Would it be better for PyPy if _seterrobj() just returned the capsule and we add it to the contextvar in Python?

I think so. Would it be less code?

mattip commented:
Thinking about naming, calling it `errstate` is confusing. Buffsize is not really an error handling parameter. If we add a `precision` field, that will make the name less appropriate. But it might be too disruptive to change it to something else like `ufuncstate`

seberg committed:
MAINT: Create extobj in C but set contextvar in python (also rename funcs)

This may be a few nanoseconds slower in CPython, but should be better
for PyPy.  It is also maybe just a bit easier to grok since the reset
of the contextvar is in Python already anyway.

seberg committed:
DOC: Add errstate changes release note and note to setbufsize.

mattip commented on doc/release/upcoming_changes/23936.improvement.rst:
```diff
@@ -0,0 +1,5 @@
+``np.errstate()`` is now faster and context safe
+------------------------------------------------
+The `np.errstate` context manager/decorator is now faster and
+safer.  Previously, it was not context save and had (rarely)
```
```suggestion
safer.  Previously, it was not context safe and had (rarely)
```

mattip commented on numpy/core/src/umath/extobj.c:
```diff
@@ -6,43 +6,73 @@
 /*
- * On return, if errobj is populated with a non-NULL value, the caller
- * owns a new reference to errobj.
+ * The global ContextVar to store the extobject this is exposed to Python
```
```suggestion
 * The global ContextVar to store the extobject. It is exposed to Python
```

mattip commented on numpy/core/src/umath/extobj.c:
```diff
@@ -6,43 +6,73 @@
- * On return, if errobj is populated with a non-NULL value, the caller
- * owns a new reference to errobj.
+ * The global ContextVar to store the extobject this is exposed to Python
+ * as `_extobj_contextvar` to reset its state in `np.errstate.__exit__`.
```
```suggestion
 * as `_extobj_contextvar`.
```

mattip reviewed (commented):
Nice. There is [a segfault](https://github.com/numpy/numpy/actions/runs/5264805314/jobs/9516791886?pr=23936#step:4:5697) (maybe at shutdown?) and some other failures I didn't look at.

I found a few nitpicky typos that can be part of the next set of commits

seberg committed:
DOC: Apply doc fixes from review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg committed:
BUG: Fix segfault due to missing incref when exposing _extobj_contextvar

seberg commented:
I aborted the timing, the full thing takes quite long nowadays.

In any case, the only thing I could think of that I expected to slow down is:
```
In [1]: a = np.ones((2, 2))
In [2]: b = np.ones(2)
In [3]: %timeit a + b
632 ns ± 0.195 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
and it actually got faster with the change: to 619ns.

When you ignore FPEs (and they get floated) the speedup is at least 50ns in total.

The `with np.errstate(4 kwargs):` overhead on my computer changes from about 2.2us to 0.7us.  Still the cost of a cheap ufunc call, but at least not 3 times it.

seberg commented:
Cirrus CI got this weird error again (going to restart, since I don't think its related)

<details>

```
f2py/tests/test_return_real.py: 16 warnings
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/f2py/tests/test_return_real.py:28: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([234], "l")) - 234.0) <= err

f2py/tests/test_return_real.py: 16 warnings
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/f2py/tests/test_return_real.py:29: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([234], "B")) - 234.0) <= err

f2py/tests/test_return_real.py: 16 warnings
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/f2py/tests/test_return_real.py:30: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([234], "f")) - 234.0) <= err

f2py/tests/test_return_real.py: 16 warnings
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/f2py/tests/test_return_real.py:31: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
    assert abs(t(array([234], "d")) - 234.0) <= err

linalg/tests/test_linalg.py::TestDet::test_types[complex64]
linalg/tests/test_linalg.py::TestDet::test_types[complex128]
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/linalg/linalg.py:2175: RuntimeWarning: divide by zero encountered in det
    r = _umath_linalg.det(a, signature=signature)

linalg/tests/test_linalg.py::TestDet::test_types[complex64]
linalg/tests/test_linalg.py::TestDet::test_types[complex128]
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/linalg/linalg.py:2175: RuntimeWarning: invalid value encountered in det
    r = _umath_linalg.det(a, signature=signature)

linalg/tests/test_linalg.py::TestDet::test_types[complex64]
linalg/tests/test_linalg.py::TestDet::test_types[complex128]
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/linalg/linalg.py:2115: RuntimeWarning: divide by zero encountered in slogdet
    sign, logdet = _umath_linalg.slogdet(a, signature=signature)

linalg/tests/test_linalg.py::TestDet::test_types[complex64]
linalg/tests/test_linalg.py::TestDet::test_types[complex128]
  /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cibw-run-tqz8ywxt/cp310-macosx_arm64/venv-test/lib/python3.10/site-packages/numpy/linalg/linalg.py:2115: RuntimeWarning: invalid value encountered in slogdet
    sign, logdet = _umath_linalg.slogdet(a, signature=signature)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED typing/tests/test_typing.py::test_code_runs[ufunc_config] - ValueError...
FAILED typing/tests/test_typing.py::test_code_runs[simple_py3] - ValueError: ...
FAILED typing/tests/test_typing.py::test_code_runs[random] - ValueError: PyCa...
FAILED typing/tests/test_typing.py::test_code_runs[index_tricks] - ValueError...
FAILED typing/tests/test_typing.py::test_code_runs[numeric] - ValueError: PyC...
FAILED typing/tests/test_typing.py::test_code_runs[simple] - ValueError: PyCa...
FAILED typing/tests/test_typing.py::test_code_runs[mod] - ValueError: PyCapsu...
FAILED typing/tests/test_typing.py::test_code_runs[arrayprint] - ValueError: ...
FAILED typing/tests/test_typing.py::test_code_runs[ufunclike] - ValueError: P...
FAILED typing/tests/test_typing.py::test_code_runs[bitwise_ops] - ValueError:...
10 failed, 35030 passed, 212 skipped, 38 xfailed, 9 xpassed, 349 warnings in 386.36s (0:06:26)
 a=
 a=

                                                                     ✕ 397.86s
Error: Command ['/bin/sh', '-c', 'bash /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build/tools/wheels/cibw_test_command.sh /private/var/folders/76/zy5ktkns50v6gt5g8r0sf6sc0000gn/T/cirrus-ci-build'] failed with code 1. None
```

seberg committed:
BUG: Fix context var setting in `seterrcall`

Impressive how few tests failed, but tests *did* fail and they
didn't segfault :)

mattip reviewed (approved):
LGTM. I will leave it a bit so others can have a look.

mattip commented:
In the previous PR there are still traces of the removed interfaces in the documentation. I see a warning when building the docs:
```
WARNING: [autosummary] failed to import numpy.geterrobj.
Possible hints:
* ModuleNotFoundError: No module named 'numpy.geterrobj'
* AttributeError: module 'numpy' has no attribute 'geterrobj'
* ImportError: 
WARNING: [autosummary] failed to import numpy.seterrobj.
```

I think that is coming from [doc/source/reference/routines.err.rst](https://github.com/numpy/numpy/blob/1d474c0aae3a550daef7f64dd422eca112f4366b/doc/source/reference/routines.err.rst?plain=1#L18)

Could you remove them in this PR or should I make another to do it?

charris commented:
Thanks Sebastian.

charris commented:
>Could you remove them in this PR or should I make another to do it?

I removed them, see #23964.
